File: 202-happy-number/202-happy-number.java
class Solution {
    public boolean isHappy(int n) {
         int slow = n;
        int fast = calSumSqr(n);
        while(fast!=1 && slow!=fast){
            slow = calSumSqr(slow);
            fast = calSumSqr(calSumSqr(fast));
        }
        return fast == 1;
    }
    private int calSumSqr(int num){
        int sumSqr = 0;
        while(num>0){
            int digit = num % 10;
            sumSqr += (digit * digit);
            num/=10;
        }
        return sumSqr;
    }
}

File: 202-happy-number/202-happy-number.cpp
class Solution {
public:
    bool isHappy(int n) {
        int sum = 0;
        int count = 0;
        int cycle = 0;
        while(true)
        {   
            sum = 0;
            while(n != 0)
            {
                sum += (n%10)*(n%10);
                n /= 10;
            }
            if(count == 10)
                cycle = sum;
            if(sum == 1)
                return true;
            else if(sum == cycle && count != 10)
                return false;
            else
                n = sum;
            count++;
        }
    }
};